### PR TITLE
GH-105: Externalize Retry

### LINKED ISSUE #105: Extract RetryTemplate to Listener Adapters

### PR DESCRIPTION
**Pull request by garyrussell**
Resolves #105
    
Extract the retry template into adapters.

### CODE REVIEW

**Comment by artembilan** [comment]
#101 has gone already.

**Comment by garyrussell** [comment]
Cool; re-pushed; thanks.

File: spring-kafka/src/main/java/org/springframework/kafka/config/AbstractKafkaListenerEndpoint.java [hunk 1]
```diff
@@ -241,24 +272,32 @@ public void setupListenerContainer(MessageListenerContainer listenerContainer, M
...
-				aml = new FilteringAcknowledgingMessageListenerAdapter<>(this.recordFilterStrategy, aml, this.ackDiscarded);
-				container.setupMessageListener(aml);
+				messageListener = new RetryingAcknowledgingMessageListenerAdapter<>(
+						(AcknowledgingMessageListener<K, V>) messageListener, this.retryTemplate,
+						this.recoveryCallback);
 			}
 			else {
-				messageListener = new FilteringMessageListenerAdapter<>(this.recordFilterStrategy, messageListener);
-				container.setupMessageListener(messageListener);
+				messageListener = new FilteringMessageListenerAdapter<>(this.recordFilterStrategy,
```
Comment (artembilan) [inline spring-kafka/src/main/java/org/springframework/kafka/config/AbstractKafkaListenerEndpoint.java]: The copy/paste hell. Must be `RetryingMessageListenerAdapter` :smile: 
Will fix on merge if only concern.

File: spring-kafka/src/main/java/org/springframework/kafka/config/AbstractKafkaListenerEndpoint.java [hunk 2]
```diff
@@ -241,24 +272,32 @@ public void setupListenerContainer(MessageListenerContainer listenerContainer, M
...
+						(MessageListener<K, V>) messageListener);
 			}
 		}
-		else {
-			container.setupMessageListener(messageListener);
+		if (this.recordFilterStrategy != null) {
+			if (messageListener instanceof AcknowledgingMessageListener) {
+				messageListener = new FilteringAcknowledgingMessageListenerAdapter<>(this.recordFilterStrategy,
+						(AcknowledgingMessageListener<K, V>) messageListener, this.ackDiscarded);
```
Comment (artembilan) [inline spring-kafka/src/main/java/org/springframework/kafka/config/AbstractKafkaListenerEndpoint.java]: Can't we revise API and really have `messageListener` as a first ctor arg?
Just for cleaner code :smile:

File: spring-kafka/src/main/java/org/springframework/kafka/listener/adapter/AbstractRetryingMessageListenerAdapter.java [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+		this.recoveryCallback = recoveryCallback != null ? recoveryCallback : new RecoveryCallback<Void>() {
+
+			@Override
+			public Void recover(RetryContext context) throws Exception {
+				@SuppressWarnings("unchecked")
+				ConsumerRecord<K, V> record = (ConsumerRecord<K, V>) context.getAttribute("record");
+				Throwable lastThrowable = context.getLastThrowable();
+				AbstractRetryingMessageListenerAdapter.this.logger.error(
+							AbstractRetryingMessageListenerAdapter.this.beanName +
+							": Listener threw an exception and retries exhausted for " + record, lastThrowable);
```
Comment (artembilan) [inline spring-kafka/src/main/java/org/springframework/kafka/listener/adapter/AbstractRetryingMessageListenerAdapter.java]: Why can't we just rethrow exception on retry exhausting by default to let `ErrorHandler` does its purpose?
If someone would like just to log an error, that is his choice.
I see your docs on the matter, but anyway would like to know the premise do it like this and make `ErrorHandler` useless together with `AbstractRetryingMessageListenerAdapter`.
Thanks

Comment (artembilan) [inline spring-kafka/src/main/java/org/springframework/kafka/listener/adapter/AbstractRetryingMessageListenerAdapter.java]: My point here is just leave `recoveryCallback` as `null` by default.

**Comment by garyrussell** [comment]
Pushed

**Comment by garyrussell** [comment]
Rebased

**Comment by artembilan** [comment]
Merged as https://github.com/spring-projects/spring-kafka/commit/e7240e207db697fd0f82de28fb95351acb8e5daa

**Pull request closed without merging**